app/uk/gov/hmrc/gform/sharedmodel/SubmissionReferenceUtil.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.sharedmodel
+import java.security.MessageDigest
+import java.math.BigInteger
+import scala.math.pow
+
+import uk.gov.hmrc.gform.sharedmodel.form.EnvelopeId
+
+object SubmissionReferenceUtil {
+
+  def findDigits(source: Long, no: Int, digits: Array[Long]): Array[Long] =
+    if (no == 0) {
+      (source % 36) +: digits
+    } else if (no == 1) {
+      findDigits(source, no - 1, ((source / 36) % 36) +: digits)
+    } else {
+      findDigits(source, no - 1, ((source / pow(36L, no)) % 36).toLong +: digits)
+    }
+
+  def getSubmissionReference(envelopeId: EnvelopeId) = {
+    val digest = MessageDigest.getInstance("SHA-256").digest(envelopeId.value.getBytes()).take(8)
```

Comment:
Unusually, it would be worthwhile to include an explanatory comment here, which states that 36^11 > 2^63 .  If it were not, then we would not be getting full significance out of the digest into our 11 digits

Reply:
Actually, I think you can use BigInt to form an integer from the array of bytes, avoiding all of this

Reply:
E.g.:

  import scala.math.BigInt
  val byteArray = BigInt(1337).toByteArray
  val int = BigInt(byteArray)